feat: public range proof

thomas-waite opened the pull request:
<!-- Please write an overview description of the problem -->
## Summary
In many use cases, it is important to be able to prove that a confidential digital asset has a value  greater than or less than a public integer. This has to be performed in zero-knowledge - that is the value of the original confidential digital asset is never revealed. We call this proof a public range proof.

This PR implements such a proof - providing the Javascript proof construction code and the smart contract proof validator, along with relevant ABI encoding functionality and tests.

## What I would like feedback on/where to focus
1) `(aztec.js/src/abiEncoder/inputCoder.js)`: To make the API for `publicRange.encodePublicRangeTransaction()` general purpose for both > and < proofs, the `inputCoder` deals with the logic required to make a < proof work. i.e. for a < proof, the flow is that a user submits a negative kPublic and then the inputCoder will behind the scenes do `(constants.GROUP_MODULUS).add(kPublic)`. Interested in thoughts on this flow and whether submitting a negative kPublic, when trying to prove that `k_1 < kPublic` feels right

2) `(protocol/contracts/ACE/validators/publicRange/PublicRangeABIEncoder.sol)`: Currently, in the bytes proofOutput object I am returning a non-zero outputNotes field - I am returning the utility note. The alternative is to return an empty field, as this output/utility note will not be used for anything after proving the greater than/less than property. Interested in thoughts on this? Returning an output note is consistent with the other proofs and function setups


<!-- What packages did you update? How were they changed? -->
## Packages Updated
- [x] aztec.js - Javascript proof construction and verification code, ABI encoding public range proof input coder and tests
- [x] protocol - zero-knowledge proof validation smart contract, ABI encoding contract and tests

<!-- Make sure that all of the points below are checked -->
## Checklist
_aztec.js_
- [x] Javascript proof construction code
- [x] Tests for well formed proof data
- [x] Javascript proof verification code
- [x] Tests for successful validation and checking of failure states
- [x] ABI input encoding 
- [x] Tests for ABI input encoding

_protocol_
- [x] zk proof validator contract
- [x] Basic tests for zk proof validator
- [x] ABI output encoding contract
- [x] Basic tests for ABI output encoding contract
- [x] Extensive tests for zk validator
- [x] Tests for case where validator is checking a less than or equal to statement

thomas-waite committed:
refactor(dev-utils): add U_TOO_BIG proof construction error code

Added for the public range proof

thomas-waite committed:
feat(aztec.js): intermittment proof construction for public range proof

thomas-waite committed:
feat(aztec.js): intermittent JS verification for public range proof

thomas-waite committed:
test(aztec.js): add tests for public range proof construction

thomas-waite committed:
test(aztec.js): add tests public range proof verification

thomas-waite committed:
feat: fix proof construction
Changed blinding factor relation to correct one

thomas-waite committed:
feat: fix public range proof verification
used correct kBar relationship

thomas-waite committed:
test: implement public range proof verification tests

thomas-waite committed:
docs(aztec.js): add JSDoc for mint and burn input coder methods

thomas-waite committed:
docs: add docs to the contract

thomas-waite committed:
docs: add docs for the contract

thomas-waite committed:
feat(aztec.js): implement ABI input encoding for public range proof

thomas-waite committed:
test(aztec.js): implement tests for public range input ABI encoding

thomas-waite committed:
refactor(aztec.js): export the public range proof from proof module

thomas-waite committed:
feat(aztec.js): implement expectedOutput result using outputCoder

thomas-waite committed:
style(aztec.js): remove unnecessary comments

thomas-waite committed:
feat(protocol): implement public range proof zk validator contract

thomas-waite committed:
feat(protocol): implement interface for zk public range proof validator

thomas-waite committed:
test(protocol): implement test for zk public range proof validator

[ci skip]

thomas-waite committed:
style(protocol): lint fixes, removing unnecessary imports

thomas-waite committed:
refactor(protocol): make zk validator contract output a boolean resukt

thomas-waite committed:
test(protocol): implement failure tests for public range proof

thomas-waite committed:
style: prettier compatible fixes throughout

thomas-waite committed:
refactor(dev-utils): remove error message U_TOO_BIG
error message is no longer needed, as refactoring the proof to allow kPublic > kmax
This is to enable the proof statement to be used for both greater than and less than cases

thomas-waite committed:
refactor: replace 'u' with 'kPublic' throughout

thomas-waite committed:
refactor(protocol): add in the pairing validation,+ doc updates

thomas-waite committed:
refactor(protocol): rename publicInteger to kPublic, call ABI encoder

Rename publicInteger variable to kPublic, in line with proof
Import the PublicRangeABIEncoder.sol contract and call upon successful completion of contract

thomas-waite committed:
feat(protocol): implement PublicRangeABIEncoder - output ABI encoding

thomas-waite committed:
feat(protocol): implement PublicRangeABIEncoderTest contract

thomas-waite committed:
test(protocol): implement test for ABI output encoding

thomas-waite committed:
refactor(protocol): link ABI encoder and zk validator, update tests

Call the ABI output encoding at the end of the public range zk validator contracts
Update the corresponding tests to expect this

thomas-waite committed:
refactor: rename publicInteger to kPublic, eslint fixes

thomas-waite committed:
test(aztec.js): update tests with new note construction API

thomas-waite committed:
test(protocol): update tests to use new note construction API

thomas-waite committed:
test(protocol): further tests checking failure cases

thomas-waite committed:
refactor(protocol): remove challenge from return

Challenge removed from return of encodePublicRangeTransaction()
Done to be consistent

thomas-waite committed:
refactor(protocol): assert n=2, so that this num of notes is required

thomas-waite committed:
refactor(aztec.js): generate blinding factors for >1 note

Needed in order to generate proof data for additional notes - to allow validator failure test for >maximum allowed num notes

thomas-waite committed:
test(protocol): validator failure test for >max num notes

apply prettier changes

thomas-waite committed:
style(aztec.js): doc updates for kPublic variable

thomas-waite committed:
feat(aztec.js): add logic to deal with kPublic < 0 case in inputCoder

thomas-waite committed:
test(aztec.js): implement kPublic < 0 test

thomas-waite committed:
style(aztec.js): eslint and prettier fixes

PaulRBerg commented on packages/aztec.js/package.json:
```diff
@@ -14,7 +14,7 @@
     "ethereumjs-util": "^6.1.0",
     "web3-eth-abi": "1.0.0-beta.50",
-    "web3-utils": "1.0.0-beta.50"
+    "web3-utils": "1.0.0-beta.52"
```
Unless your other changes depend on this version of the utils, I would offload these updates onto @dependabot-bot.

thomas-waite replied:
Good point

thomas-waite committed:
refactor(aztec.js): fix rebase problems

ArnSch commented on packages/aztec.js/src/proof/publicRange/index.js:
```diff
@@ -0,0 +1,180 @@
+ * @memberof module:publicRange
+ * @param {Note[]} inputNotes input AZTEC notes
+ * @param {Note[]} outputNotes output AZTEC notes
+ * @param {Number} kPublic - public integer being compared against
```
I'm not a fan of reusing kPublic (which is an unclear variable name already in joinSplit). It seems like it will just add confusion as to what kPublic is. Can we not name it `publicComparaison` or something similar?

thomas-waite replied:
Sure, done

ArnSch commented on packages/aztec.js/src/proof/publicRange/index.js:
```diff
@@ -0,0 +1,180 @@
+ * @param {string} senderAddress the Ethereum address sending the AZTEC transaction (not necessarily the note signer)
+ * @returns {Object} AZTEC proof data and expected output
+ */
+publicRange.encodePublicRangeTransaction = ({ inputNotes, outputNotes, kPublic, senderAddress }) => {
```
Why are we constraining ourselves to this really un-descriptive API? In this context, I wouldn't expect anyone to have any clue what inputNotes, outputNotes, or kPublic are supposed to be. 

Private range, for example, is much more descriptive. If we go in that direction, we can also feed in a boolean to denote whether this is a < or a > range comparison.

thomas-waite replied:
Agreed, changed the interface to:

```
publicRange.encodePublicRangeTransaction = async ({
    originalNote,
    publicComparison,
    senderAddress,
    isGreaterOrEqual,
    utilityNote,
}) => { }
```

The utilityNote is an optional parameter - the default is that it's generated internally and is not required as an input - but some flows require it to be possible to explicitly input

ArnSch commented on packages/aztec.js/src/proof/publicRange/index.js:
```diff
@@ -0,0 +1,180 @@
+ *
+ * @method constructBlindingFactors
+ * @param {Object[]} notes AZTEC notes
+ * @param {Number} kPublicBN - BN instance of the public integer being compared against
```
Misleading documentation

thomas-waite replied:
Yep, fixed

ArnSch commented on packages/aztec.js/test/abiEncoder/encoderFactory.js:
```diff
@@ -453,4 +449,66 @@ describe('abiEncoder.encoderFactory', () => {
+            // Setup
+            let accounts = [];
+            let notes = [];
+            const u = 10;
```
u?

thomas-waite replied:
Done

ArnSch commented on packages/aztec.js/test/proof/publicRange/verifier.spec.js:
```diff
@@ -0,0 +1,154 @@
+
+describe('Public range proof verifier', () => {
+    describe('Success States', () => {
+        it('should construct a valid dividend computation proof', async () => {
```
Invalid test name.

thomas-waite replied:
Done

ArnSch commented on packages/protocol/contracts/ACE/validators/publicRange/PublicRange.sol:
```diff
@@ -0,0 +1,325 @@
+import "./PublicRangeABIEncoder.sol";
+
+/**
+ * @title Library to validate AZTEC Bilateral Swap zero-knowledge proofs
```
Please update the documentation.

thomas-waite replied:
Done

ArnSch commented on packages/protocol/contracts/ACE/validators/publicRange/PublicRange.sol:
```diff
@@ -0,0 +1,325 @@
+    /**
+     * @dev PublicRange.sol will take any transaction sent to it and attempt to validate a zero knowledge proof.
+     * If the proof is not valid, the transaction throws.
+     * @notice See BilateralSwapInterface for how method calls should be constructed.
```
See above.

thomas-waite replied:
Done

ArnSch commented on packages/protocol/contracts/ACE/validators/publicRange/PublicRange.sol:
```diff
@@ -0,0 +1,325 @@
+            // We don't check for function signatures, there's only one function that 
+            // ever gets called: validateBilateralSwap()
+            // We still assume calldata is offset by 4 bytes so that we can represent 
+            // this contract through a comp\atible ABI
```
Typo

thomas-waite replied:
Done

ArnSch reviewed (changes requested):
I don't think this is good to go, mostly because this code contains quite a few copy-paste errors, and because the API to proof construction for this proof isn't very good.

To answer your two questions specifically:

1) I think this whole interface need uphauling.
2) I lean slightly in the direction of not returning data just for the sake of it, but up to you.

thomas-waite committed:
feat: add missing proofIDs

thomas-waite committed:
style: rename kPublic to publicComparison

thomas-waite committed:
docs: update docs to accurately reflect the publicRange proof

thomas-waite committed:
refactor(aztec.js): implement PR feedback, make API more intuitive

construct utility note internally inside the encodePublicRangeTransaction() function. Create the helpers method constructUtilityNote() to do this
remove the inputNotes and outputNotes arguments
add a boolean to govern whether this proof for greater than or less than
add a check that publicComparison integer is well formed
make it possible, but optional as to whether a utility note can be manually inputted

thomas-waite committed:
fix(protocol): return publicValue as 0, rather than publicComparison

thomas-waite committed:
refactor: add kPublic, publicOwner into hash used to compute challenge

done to make challenge more unique. kPublic = 0 and publicOwner = 0x00...0 for this proof

thomas-waite committed:
fix(aztec.js): bug fix in automatic creation of utility note

in the case where originalValue === comparisonValue, the utilityValue was being set to comparisonValue rather than 0

thomas-waite committed:
fix(aztec.js): bug fix in automatic creation of utility note

in the case where originalValue === comparisonValue, the utilityValue was being set to comparisonValue rather than 0
helper function to check if integer

thomas-waite committed:
test: update all tests with new encodePublicRangeTransaction() API

thomas-waite committed:
style: make tests setup more intuitive

remove noteValues array and explicitly assign originalNoteValue and utilityNoteValue
update docs

thomas-waite committed:
feat(aztec.js): add all notes into range proof

make use of batch bilinear pairing optimisation

thomas-waite committed:
feat(protocol): add all notes into bilinear pairing check

update the validator contracts
add docs

thomas-waite committed:
test(protocol): check fail for wrong proof inequality

supply a less than proof, but specify that it is a greater than proof, and check failure
update test names to follow standardised format
apply prettier
